flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,738 @@
+    bin_cols_without_df_name = [c for c in bin_cols_in if c != "df_name"]
+    bin_cols_with_df_name = bin_cols_in + ["df_name"]
+
+    # Add the wd and ws bins
+    df_ = _add_wd_ws_bins(
+        df_,
+        wd_cols=wd_cols,
+        ws_cols=ws_cols,
+        wd_step=wd_step,
+        wd_min=wd_min,
+        wd_max=wd_max,
+        ws_step=ws_step,
+        ws_min=ws_min,
+        ws_max=ws_max,
+    )
+
+    # Compute the covariance frame
+    df_cov = _compute_covariance(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    # If filling missing covariance terms, do it now
+    if fill_cov_with_var:
+        df_cov = _fill_cov_null(df_cov, test_cols=test_cols)
+
+    # If only using the variance, zero out the covariance terms
+    if variance_only:
+        df_cov = _zero_cov(df_cov, test_cols=test_cols)
+
+    # Apply Null values to covariance and sync across uplift pairs
+    df_cov = _null_and_sync_covariance(
+        df_cov=df_cov,
+        test_cols=test_cols,
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_cov)
+
+    # Bin and group the dataframe
+    df_bin = _bin_and_group_dataframe_expected_power(
+        df_=df_,
+        test_cols=test_cols,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # Synchronize the null values
+    df_bin = _synchronize_nulls(
+        df_bin=df_bin,
+        sync_cols=[f"{col}_mean" for col in test_cols],
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # Remove rows where any of the test columns are null (if remove_any_null_turbine_bins is True)
+    if remove_any_null_turbine_bins:
+        df_bin = df_bin.filter(
+            pl.all_horizontal([pl.col(f"{c}_mean").is_not_null() for c in test_cols])
+        )
+
+    # Join the covariance dataframe to df_bin
+    df_bin = df_bin.join(df_cov, on=bin_cols_with_df_name, how="left")
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+
+    # Get the names of all the covariance and num_points columns
+    cov_cols = [f"cov_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+    n_cols = [f"count_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+
+    # Compute the farm power
+    df_bin = df_bin.with_columns(pow_farm=pl.sum_horizontal([f"{c}_mean" for c in test_cols]))
+
+    # Computed the expected farm power variance as the sum of cov_cols / n_cols
+    # This is equation 4.16
+    df_bin = df_bin.with_columns(
+        pow_farm_var=pl.sum_horizontal(
+            [pl.col(cov) / pl.col(n) for cov, n in zip(cov_cols, n_cols)]
+        )
+    )
+
+    # If any of the cov_cols are null, set pow_farm_var to null
+    df_bin = df_bin.with_columns(
+        pl.when(pl.all_horizontal([pl.col(c).is_not_null() for c in cov_cols]))
```

Comment:
Connected to the comment above, I think we could get rid of this requirement that all cov_cols are defined. When computing the "pow_farm_var" column a few lines above, if my understanding is correct, the summation over cov_cols will just ignore any Null values. So, we'd be calculating the variance of the farm power considering only the turbines that are valid, in the same way that "pow_farm" will only sum the power of the turbines that are not Null. Therefore, the farm power variance will correspond to the same set of turbiens used to compute farm power.

Reply:
This holds now I think so commenting out